plugins/org.eclipse.n4js.flowgraphs/src/org/eclipse/n4js/flowgraphs/analysers/DeadCodeFinder.java

```diff
@@ -0,0 +1,280 @@
+import org.eclipse.n4js.n4JS.Block;
+import org.eclipse.n4js.n4JS.ControlFlowElement;
+import org.eclipse.n4js.n4JS.ExpressionStatement;
+import org.eclipse.n4js.n4JS.FunctionOrFieldAccessor;
+import org.eclipse.n4js.n4JS.ReturnStatement;
+import org.eclipse.n4js.n4JS.Statement;
+import org.eclipse.xtext.EcoreUtil2;
+import org.eclipse.xtext.nodemodel.ICompositeNode;
+import org.eclipse.xtext.nodemodel.util.NodeModelUtils;
+import org.eclipse.xtext.util.TextRegion;
+
+/**
+ * Collects all reachable nodes and hence finds all unreachable nodes, alias <i>dead code</i>. Provides methods to
+ * determine if a given {@link ControlFlowElement} is dead code, or to compute a minimal set of {@link TextRegion}s of
+ * dead code.
+ */
+public class DeadCodeFinder extends GraphVisitor {
+	Set<ControlFlowElement> allForwardCFEs = new HashSet<>();
+	Set<ControlFlowElement> allBackwardCFEs = new HashSet<>();
+	Set<ControlFlowElement> allIslandsCFEs = new HashSet<>();
+	Set<ControlFlowElement> allCatchBlocksCFEs = new HashSet<>();
+	Set<ControlFlowElement> unreachableCFEs = new HashSet<>();
+
+	/** Constructor */
+	public DeadCodeFinder() {
+		super(Direction.Forward, Direction.Backward, Direction.Islands, Direction.CatchBlocks);
+	}
+
+	@Override
+	protected void initAll() {
+		// nothing to do
+	}
+
+	@Override
+	protected void init(Direction curDirection, ControlFlowElement curContainer) {
+		// nothing to do
+	}
+
+	@Override
+	protected void visit(ControlFlowElement cfe) {
+		switch (getCurrentDirection()) {
+		case Forward:
+			allForwardCFEs.add(cfe);
+			break;
+		case Backward:
+			allBackwardCFEs.add(cfe);
+			break;
+		case Islands:
+			allIslandsCFEs.add(cfe);
+			break;
+		case CatchBlocks:
+			allCatchBlocksCFEs.add(cfe);
+			break;
+		}
+	}
+
+	@Override
+	protected void visit(ControlFlowElement start, ControlFlowElement end, FlowEdge edge) {
+		// nothing to do
+	}
+
+	@Override
+	protected void terminate(Direction curDirection, ControlFlowElement curContainer) {
+		// nothing to do
+	}
+
+	@Override
+	protected void terminateAll() {
+		unreachableCFEs.addAll(allBackwardCFEs);
+		unreachableCFEs.removeAll(allForwardCFEs);
+		unreachableCFEs.addAll(allIslandsCFEs);
+		unreachableCFEs.removeAll(allCatchBlocksCFEs);
+	}
+
+	/** @return all reachable {@link ControlFlowElement}s */
+	public Set<ControlFlowElement> getReachableCFEs() {
+		return allForwardCFEs;
+	}
+
+	/** @return all unreachable {@link ControlFlowElement}s */
+	public Set<ControlFlowElement> getUnreachableCFEs() {
+		return unreachableCFEs;
+	}
+
+	/**
+	 * This method deals with the fact that {@link Statement}s are not represented in the control flow graph.
+	 *
+	 * @return true iff the given {@link ControlFlowElement} is dead code.
+	 */
+	public boolean isDeadCode(ControlFlowElement cfe) {
+		cfe = CFEMapper.map(cfe);
+
+		if (FGUtils.isControlElement(cfe)) {
+			Set<ControlFlowElement> succs = flowAnalyses.getSuccessors(cfe);
+			for (ControlFlowElement succ : succs) {
```

Comment:
Similar to @yoosiba , I also notice that many times instead of using `functional style`, the code uses a normal for loop to filter elements/search for elements. But I think @mmews-n4 intentionally uses this "old school" style for readability?

Reply:
Yes. I avoid streams on purpose since I don't like to debug them.